chemprop/uncertainty/calibrator.py

```diff
@@ -78,58 +81,251 @@ def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
+                )
+            qhat = torch.quantile(scores_j, q_level, interpolation="higher")
+            self.qhats.append(qhat)
+
+        self.qhats = torch.tensor(self.qhats)
 
     def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
-        ...
-        return
+        calibrated_preds = torch.zeros_like(uncs, dtype=torch.int)
+        scores = self.nonconformity_scores(uncs)
+
+        for j, qhat in enumerate(self.qhats):
+            calibrated_preds[:, j] = (scores[:, j] <= qhat).int()
+
+        return preds, calibrated_preds
 
 
 @UncertaintyCalibratorRegistry.register("conformal-adaptive")
-class ConformalAdaptiveMulticlassCalibrator(UncertaintyCalibrator):
-    def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> None:
-        ...
-        return
+class AdaptiveMulticlassConformalCalibrator(MulticlassConformalCalibrator):
+    @staticmethod
+    def nonconformity_scores(preds):
+        r"""Compute nonconformity score by greedily including classes in the classification set until it reach the true label.
 
-    def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
-        ...
-        return
+        .. math::
+            s(x, y) = \sum_{j=1}^{k} \hat{f}(x)_{\pi_j(x)}, \text{ where } y = \pi_k(x)
 
+        where :math:`\pi_k(x)` is the permutation of :math:`\{1,...,K\}` that sorts :math:`\hat{f}(X_{test})` from most likely to least likely.
+        """
 
-@UncertaintyCalibratorRegistry.register("conformal-regression")
-class ConformalRegressionCalibrator(UncertaintyCalibrator):
-    def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> None:
-        ...
-        return
+        sort_index = torch.argsort(-preds, dim=2)
+        sorted_preds = torch.gather(preds, 2, sort_index)
+        sorted_scores = sorted_preds.cumsum(dim=2)
+        unsorted_scores = torch.zeros_like(sorted_scores).scatter_(2, sort_index, sorted_scores)
 
-    def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
-        ...
-        return
+        return unsorted_scores
 
 
-@UncertaintyCalibratorRegistry.register("conformal-quantile-regression")
-class ConformalQuantileRegressionCalibrator(UncertaintyCalibrator):
+@UncertaintyCalibratorRegistry.register("conformal-regression")
+class RegressionConformalCalibrator(UncertaintyCalibrator):
+    r"""Conformalize quantiles to make the interval :math:`[\hat{t}_{\alpha/2}(x),\hat{t}_{1-\alpha/2}(x)]` to have
+    approximately 1 - :math:`\alpha` coverage. [1]_
+
+    .. math::
+        s(x, y) &= \max \left\{ \hat{t}_{\alpha/2}(x) - y, y - \hat{t}_{1-\alpha/2}(x) \right\}
+
+        \hat{q} &= \text{Quantile}(s_1, \ldots, s_n; \left\lceil \frac{(n+1)(1-\alpha)}{n} \right\rceil)
+
+        C(x) &= \left[ \hat{t}_{\alpha/2}(x) - \hat{q}, \hat{t}_{1-\alpha/2}(x) + \hat{q} \right]
```

Comment:
do blank lines work for `align` environments instead of manual line breaks (`\\`)?

Reply:
I think it works based on the documentation it creates (https://chemprop--989.org.readthedocs.build/en/989/autoapi/chemprop/uncertainty/calibrator/index.html#chemprop.uncertainty.calibrator.RegressionConformalCalibrator).
<img width="805" alt="Screen Shot 2024-08-26 at 1 12 33 PM" src="https://github.com/user-attachments/assets/306e0076-23c3-4230-ad06-d68572cabeca">